session.go

```diff
@@ -11,20 +16,89 @@ import (
 type Session struct {
 	sql.Session
 	Pool *RepositoryPool
+
+	bblfshMu       sync.Mutex
+	bblfshEndpoint string
+	bblfshClient   *bblfsh.Client
+}
+
+// Config values for a GitQuery session.
+type Config struct {
+	Pool           *RepositoryPool
+	BblfshEndpoint string
 }
 
+const (
+	bblfshEndpointKey     = "BBLFSH_ENDPOINT"
+	defaultBblfshEndpoint = "127.0.0.1:9432"
+)
+
 // NewSession creates a new Session.
-func NewSession(pool *RepositoryPool) *Session {
+func NewSession(conf *Config) *Session {
+	if conf.BblfshEndpoint == "" {
+		conf.BblfshEndpoint = getStringEnv(bblfshEndpointKey, defaultBblfshEndpoint)
+	}
+
 	return &Session{
-		Session: sql.NewBaseSession(),
-		Pool:    pool,
+		Session:        sql.NewBaseSession(),
+		Pool:           conf.Pool,
+		bblfshEndpoint: conf.BblfshEndpoint,
+	}
+}
+
+const bblfshMaxAttempts = 10
+
+// BblfshClient returns a BblfshClient.
+func (s *Session) BblfshClient() (*bblfsh.Client, error) {
+	var err error
+	s.bblfshMu.Lock()
+	defer s.bblfshMu.Unlock()
+
+	if s.bblfshClient == nil {
+		s.bblfshClient, err = bblfsh.NewClient(s.bblfshEndpoint)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var attempts int
+	for {
+		if attempts > bblfshMaxAttempts {
+			return nil, ErrBblfshConnection.New()
+		}
+
+		switch s.bblfshClient.GetState() {
+		case connectivity.Ready, connectivity.Idle:
+			return s.bblfshClient, nil
+		case connectivity.Connecting:
+			time.Sleep(100 * time.Millisecond)
```

Comment:
Shouldn't we reset or decrease attempt counter? If it's connection at attempt 10 it will exit the next loop.

Reply:
That could lead to it being in an infinite loop if it's connecting, then fails, then connecting, then fails. We could set a maximum total retries and reset only a local counter, tho